src/SNLS_device_forall.h

```diff
@@ -50,10 +56,99 @@ end,                                             \
 
 namespace snls {
 
-   typedef RAJA::View<bool, RAJA::Layout<1> > rview1b;
-   typedef RAJA::View<double, RAJA::Layout<1> > rview1d;
-   typedef RAJA::View<double, RAJA::Layout<2> > rview2d;
-   typedef RAJA::View<double, RAJA::Layout<3> > rview3d;
+   // Provide some simple shortcuts in-case people need something beyond the default
+   template<typename T>
+   using rview1 = RAJA::View<T, RAJA::Layout<1>>;
+   template<typename T>
+   using rview2 = RAJA::View<T, RAJA::Layout<2>>;
+   template<typename T>
+   using rview3 = RAJA::View<T, RAJA::Layout<3>>;
+
+   using rview1b = rview1<bool>;
+   using rview1d = rview1<double>;
+   using rview2d = rview2<double>;
+   using rview3d = rview3<double>;
+   using crview1d = rview1<const double>;
+   using crview2d = rview2<const double>;
+   using crview3d = rview3<const double>;
+
+   using rhost_res = RAJA::resources::Host;
+#if defined(__snls_gpu_active__)
+#if defined(RAJA_ENABLE_CUDA)
+         using rgpu_res = RAJA::resources::Cuda;
+#else //defined(RAJA_ENABLE_HIP)
+         using rgpu_res = RAJA::resources::Hip;
+#endif
+#endif
+
+   using rres = std::variant<
+      rhost_res
+#if defined(__snls_gpu_active__)
+      ,
+      rgpu_res
+#endif
+   >;
+
+   using rrese = RAJA::resources::Event;
+
+   // We really don't care what View class we're using as the sub-view just wraps it up
+   // and then allows us to take a slice/window of the original view
+   // Should probably work out some form of SFINAE to ensure T that we're templated on
+   // is an actual View that we can use.
+   template<class T>
+   class subview {
+   public:
+      // Delete the default constructor as that wouldn't be a valid object
+      __snls_hdev__
+      subview() = delete;
+      // where we don't want any offset within the subview itself
+      __snls_hdev__
+      subview(const int index, T& view) : m_view(view), m_index(index), m_offset(0) {};
+      // sometimes you might want to have an initial offset in your subview when constructing
+      // your subview in which everything appears as 0 afterwards
+      __snls_hdev__
+      subview(const int index, const size_t offset, T& view) : m_view(view), m_index(index), m_offset(offset) {};
+
+      ~subview() = default;
+
+      // Could probably add default copy constructors as well here if need be...
+
+      // Let the compiler figure out the correct return type here as the one from
+      // RAJA at least for regular Views is non-trivial
+      // make the assumption here that we're using row-major memory order for views
+      // so m_index is in the location of the slowest moving index as this is the default
+      // for RAJA...
+      template <typename... Args>
+      __snls_hdev__
+      inline
+      constexpr
```

Comment:
Did you intentionally make this constexpr?  It's not a bad idea, but it would only make sense if the object is constexpr constructable, which it isn't right now.  Perhaps we should make the constructors constexpr as well?

Reply:
Ahh nevermind, we don't need to have a constexpr constructor.  But it still could help performance if we do add it in.

Reply:
Forgot to reply here but the `constexpr` aspect if I recall correctly I borrowed from how `RAJA::Views` do this same operation.